### ISSUE TITLE
[Feature Request] Current best configuration based on the trained model

### ISSUE THREAD

**Issue by Strange369** [pytorch/botorch#671]
# 🚀 Feature Request

When we use Baysian Optimization to tune the hyper-parameters of a machine learning algorithm, we are trying to optimize the hyper-parameters in order to maximize the accuracy of prediction.
To speed up the tunning process, multiple sets of hyper-parameters are evaluated simultaneously. Therefore, I refer to the tutorial of q-EI/q-NEI (https://botorch.org/tutorials/closed_loop_botorch_only). 
In the tutorial, the final result is only the current best objective, i.e., the maximum observed value of neg_hartmann6. It is relatively easy to get the current best configuration, i.e., the corresponding train_x for the best_observed result.

However, once the surrogate model is trained, it should have the ability to generate an optimized configuration, which is assumed to result in a best result of the objective (even assumed to be better than the best_observed). One example is in Ax (https://ax.dev/tutorials/tune_cnn.html). 
Since I also want to use the feature of q-EI and q-NEI, the simple example in Ax may not be suitable for my utilization.

Therefore, How can I do that based on the tutorial of q-EI and q-NEI? 
Thank you very much for your time :)

**Comment by saitcakmak** [comment]
> once the surrogate model is trained, it should have the ability to generate an optimized configuration,

If by this you refer to the maximizer of the posterior mean, BoTorch provides a `PosteriorMean` acquisition function, which can be optimized with `optimize_acqf` to get the current best predicted solution and the corresponding predicted posterior mean value. You would simply construct this as `pm = PosteriorMean(model)` and call `optimize_acqf` as in the tutorial.

**Comment by Strange369** [comment]
I also want to use q-EI and q-NEI to generate multiple candidates for evaluating in the next iteration. If the PosteriorMean is applied as the acquision function, how can I combine them?

**Comment by Balandat** [comment]
So you don't really want to use the `PosteriorMean` to generate the points during the optimization. You would take all the data that you've observed over the optimization, then fit a model, and then use `PosteriorMean` to get an optimized candidate. So although `PosteriorMean` is an `AcquisitionFunction`, you'd not really be using it as such, you'd be using it as an "ExploitationFunction" if that makes sense.

**Comment by saitcakmak** [comment]
> I also want to use q-EI and q-NEI to generate multiple candidates for evaluating in the next iteration.

To generate multiple candidates, you specify the `q` argument to `optimize_acqf`. `q` is the number of candidates it will produce for parallel evaluation. In the tutorial you linked, this is done in Cell 5, by setting `q=BATCH_SIZE` where `BATCH_SIZE=3`. With this, `optimize_acqf` would return a `q x d`-dim tensor of solutions which would represent the `q` candidates the acquisition function suggests for evaluation.

> If the PosteriorMean is applied as the acquision function, how can I combine them?

`PosteriorMean` is only for getting the maximizer of the posterior mean, i.e., the best predicted point. You should not use it to guide function evaluation. After you're done with the BO loop using some other acquisition function, you can use it to find the predicted best point, and use that as your best configuration if you choose to.

**Comment by Strange369** [comment]
Aha, now I understand what you mean.
So, I still use the tutorial of q-EI to process the BO loop. Once I obtained all the [train_x, obj_x], feed them to PosteriorMean to get the final prediction for best point and its configuration.
This idea should work. I will have a try.
Thank you two very much!

**Comment by Balandat** [comment]
Just make sure you can trust the model (e.g. by evaluating the cross-validation performance). If the model isn't great it can still be helpful in guiding exploration, but may be problematic for determining configurations to actually launch ("out of sample best point selection"). Using the best among the observed points is the safest bet if your observations are noiseless. If your observations are noisy, using the model to make in-sample predictions at the observed points can also be valuable to regularize and shrink observations towards the mean.
